### PR TITLE
[GOBBLIN-1547] Set completion watermark to end of hour instead of beg…

### PR DESCRIPTION
**Pull request by vikrambohra**
…inning

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-1547] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1547
    - original ticket: https://issues.apache.org/jira/browse/GOBBLIN-1540

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
Details in the ticket https://issues.apache.org/jira/browse/GOBBLIN-1547

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java [hunk 1]
```diff
@@ -729,9 +729,9 @@ private StructLike addLatePartitionValueToIcebergTable(Table table, TableMetadat
   private int isLate(String datepartition, long previousWatermark) {
     ZonedDateTime partitionDateTime = ZonedDateTime.parse(datepartition, HOURLY_DATEPARTITION_FORMAT);
     long partitionEpochTime = partitionDateTime.toInstant().toEpochMilli();
-    if(partitionEpochTime > previousWatermark) {
+    if(partitionEpochTime >= previousWatermark) {
```
Comment (ZihanLi58) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java:732]: Why do we change this part? If current watermark is 09am, does that mean data for 09 has completed? And then if we see dat for 09am, is that a late data?

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]
one minor comment

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java (hunk 1 above)
Comment (vikrambohra) [reply gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java:732]: if CW = 9am then data for datepartition <= 8am is complete and datepartion>=9 is still not complete (so late =0)

**Review by ZihanLi58 — APPROVED** [review]
+1

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java [hunk 2]
```diff
@@ -861,11 +861,13 @@ public void flush(String dbName, String tableName) throws IOException {
   }
 
   /**
+   * NOTE: cw for a window [t1, t2] is marked as t2 if audit counts match
```
Comment (autumnust) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java:864]: Let's avoid abbr. like `cw` in the comments.

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java [hunk 3]
```diff
@@ -729,9 +729,9 @@ private StructLike addLatePartitionValueToIcebergTable(Table table, TableMetadat
   private int isLate(String datepartition, long previousWatermark) {
     ZonedDateTime partitionDateTime = ZonedDateTime.parse(datepartition, HOURLY_DATEPARTITION_FORMAT);
     long partitionEpochTime = partitionDateTime.toInstant().toEpochMilli();
-    if(partitionEpochTime > previousWatermark) {
+    if(partitionEpochTime >= previousWatermark) {
       return 0;
-    } else if(partitionEpochTime <= previousWatermark && partitionDateTime.toLocalDate().equals(getDateFromEpochMillis(previousWatermark))) {
+    } else if(partitionEpochTime < previousWatermark && partitionDateTime.toLocalDate().equals(getDateFromEpochMillis(previousWatermark))) {
```
Comment (autumnust) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java:734]: this line seems way too long to me, did you apply style file in your IDE ?

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java [hunk 4]
```diff
@@ -895,9 +897,9 @@ private long computeCompletenessWatermark(String table, SortedSet<ZonedDateTime>
       while (iterator.hasNext()) {
         ZonedDateTime timestampDT = iterator.next();
         if (timestampDT.isAfter(prevWatermarkDT)
-            && TimeIterator.durationBetween(prevWatermarkDT, now, granularity) > 1) {
+            && TimeIterator.durationBetween(prevWatermarkDT, now, granularity) > 0) {
           long timestampMillis = timestampDT.toInstant().toEpochMilli();
-          if(auditCountVerifier.get().isComplete(table, timestampMillis, TimeIterator.inc(timestampDT, granularity, 1).toInstant().toEpochMilli())) {
+          if(auditCountVerifier.get().isComplete(table, TimeIterator.dec(timestampDT, granularity, 1).toInstant().toEpochMilli(), timestampMillis)) {
```
Comment (autumnust) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java:902]: there's no blank after `(` and line is too long

**Review by autumnust — CHANGES_REQUESTED** [review]
As a general comments, please also add more description about this PR, thanks

**Review by autumnust — APPROVED** [review]
+1

**Pull request merged**